toolchain/check/context.cpp

```diff
@@ -233,8 +230,7 @@ auto Context::AddConvergenceBlockWithArgAndPush(
   // Acquire the result value.
   SemIR::TypeId result_type_id =
       semantics_ir().GetNode(*block_args.begin()).type_id();
-  return AddNode(
-      SemIR::Node::BlockArg::Make(parse_node, result_type_id, new_block_id));
+  return AddNode(SemIR::BlockArg(parse_node, result_type_id, new_block_id));
```

Comment:
I think if we are worried about getting the order wrong when reading out of a SemIR Node, we arguably should also be worried about getting the order wrong when constructing one. Can we talks about some factory function options that associate names with arguments?

Reply:
It'd be easy to make this work:

```suggestion
  return AddNode(SemIR::BlockArg(parse_node, result_type_id, {.block_id = new_block_id}));
```

Or this:

```suggestion
  return AddNode(SemIR::BlockArg({.parse_node = parse_node}, {.type_id = result_type_id},
                                 {.block_id = new_block_id}));
```

If we want a single braced list with the `parse_node`, `type_id`, and node operands, that's going to require more reworking of how we construct the node types.

Reply:
FWIW, I'm much less concerned. I think we'll have many more places inspecting fields than constructing nodes, and I think the code querying fields is more at risk of silently "working" in ways that are confusing.

If this is a major issue, I'd hope to try making the parameters more type-safe first.

Reply:
Here's what that might look like: https://github.com/carbon-language/carbon-lang/commit/5a2e244270c7416d32af3e1c42c198bb05d50037 (just a sample: not all uses of the constructor have been updated).

Interestingly, this immediately found some cases where nodes had inappropriate names for their operands.

Reply:
(my comment above still holds for me at least after seeing the result)

Reply:
In discussion, I agreed to postpone doing this until we observe there to be problems it would avoid.